source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+    bbq,
+    kchannel::{KProducer, LeakedKProducer},
+    rosc::Sender,
+};
+
+/// A partial list of known UUIDs of driver services
+pub static KNOWN_UUIDS: &[Uuid] = &[
+    //
+    // Kernel UUIDs
+    //
+
+    // SerialMux
+    uuid!("54c983fa-736f-4223-b90d-c4360a308647"),
+    // Simple Serial Port
+    uuid!("f06aac01-2773-4266-8681-583ffe756554"),
+];
+
+/// A marker trait designating a registerable driver service.
+///
+/// Typically used with [Registry::set] or [Registry::set_konly].
+/// Can typically be retrieved by [Registry::get] or [Registry::get_userspace]
+/// After the service has been registered.
+pub trait RegisteredDriver {
+    /// This is the type of the request sent TO the driver service
+    type Request: 'static;
+
+    /// This is the type of the response sent FROM the driver service
+    type Response: 'static;
+
+    /// This is the UUID of the driver service
+    const UUID: Uuid;
+
+    /// Get the type_id used to make sure that driver instances are correctly typed.
+    /// Corresponds to the same type ID as `(Self::Request, Self::Response)`
+    fn type_id() -> TypeId {
+        TypeId::of::<(Self::Request, Self::Response)>()
+    }
+}
+
+/// The driver registry used by the kernel.
+pub struct Registry {
+    items: HeapFixedVec<RegistryItem>,
+}
+
+// TODO: This probably goes into the ABI crate, here is fine for now
+#[derive(Serialize, Deserialize)]
+pub struct UserRequest<'a> {
+    // TODO: Maybe not the UUID, maybe pre-discover a shorter UID?
+    uid: Uuid,
+    nonce: u32,
+    #[serde(borrow)]
+    req_bytes: &'a [u8],
+}
+
+// TODO: This probably goes into the ABI crate, here is fine for now
+#[derive(Serialize, Deserialize)]
+pub struct UserResponse<U> {
+    // TODO: Maybe not the UUID, maybe pre-discover a shorter UID?
+    uuid: Uuid,
+    nonce: u32,
+    reply: Result<U, ()>,
+}
+
+/// A wrapper for a message TO and FROM a driver service.
+/// Used to be able to add additional message metadata without
+/// changing the fundamental message type.
+pub struct Envelope<P> {
+    pub body: P,
+}
+
+/// The [Message] kind represents a full reply/response sequence to
+/// a driver service. This is the concrete type received by the driver
+/// service.
+///
+/// It contains the Request, e.g. [RegisteredDriver::Request], as well
+/// as a [ReplyTo] that allows the driver service to respond to a given
+/// request
+pub struct Message<RD: RegisteredDriver> {
+    pub msg: Envelope<RD::Request>,
+    pub reply: ReplyTo<RD::Response>,
+}
+
+/// A `ReplyTo` is used to allow the CLIENT of a service to choose the
+/// way that the driver SERVICE replies to us. Essentially, this acts
+/// as a "self addressed stamped envelope" for the SERVICE to use to
+/// reply to the CLIENT.
+pub enum ReplyTo<U> {
+    // This can be used to reply directly to another kernel entity,
+    // without a serialization step
+    KChannel(KProducer<Envelope<Result<U, ()>>>),
+
+    Rosc(Sender<Envelope<Result<U, ()>>>),
```

Comment:
is it possible to send a `Rosc` reply to userspace, or is that kernel only as well?

Reply:
Rosc is kernel only, I'll update the comment.